### ISSUE TITLE
Cannot reproduce the paper result and checkpoint result

### ISSUE THREAD

**Issue by hyunjunChhoi** [tianyu0207/PEBAL#24]
Hello , Thanks for your work

I followed the pebal config and trained using batch size 16 (as the paper) and 40 epochs
However I cannot reproduce the paper result and check point result
Is there any tips for reproducing the high AUPRC on the FS LS dataset??
Or is there any way to reproduce the result that has similar performance with checkpoint 
Could you help??

Thanks in advance
Config :
import os
import numpy
from easydict import EasyDict

C = EasyDict()
config = C
cfg = C

C.seed = 666

"""Root Directory Config"""
C.repo_name = 'pebal'
C.root_dir = os.path.realpath(".")

"""Data Dir and Weight Dir"""
C.city_root_path = '/home/numb7315/PEBAL/code/dataset/city_scape'  # path/to/your/city_scape
C.coco_root_path = '/home/numb7315/PEBAL/code/dataset/coco'  # path/to/your/coco
C.fishy_root_path = '/home/numb7315/PEBAL/code/dataset/fishyscapes'  # path/to/your/fishy

C.pebal_weight_path = os.path.join(C.root_dir, 'ckpts', 'pebal', 'best_ad_ckpt.pth')
C.pretrained_weight_path = os.path.join(C.root_dir, 'ckpts', 'pretrained_ckpts', 'cityscapes_best.pth')

"""Network Config"""
C.fix_bias = True
C.bn_eps = 1e-5
C.bn_momentum = 0.1

"""Image Config"""
C.num_classes = 19+1   # NOTE: 1 more channel for gambler loss
C.image_mean = numpy.array([0.485, 0.456, 0.406])  # 0.485, 0.456, 0.406
C.image_std = numpy.array([0.229, 0.224, 0.225])

C.image_height = 900
C.image_width = 900

C.num_train_imgs = 2975
C.num_eval_imgs = 500

"""Train Config"""
C.lr = 1e-5
C.batch_size = 8
C.lr_power = 0.9
C.momentum = 0.9
C.weight_decay = 1e-4

C.nepochs = 40
C.niters_per_epoch = C.num_train_imgs // C.batch_size
C.num_workers = 8
C.train_scale_array = [0.5, 0.75, 1, 1.5, 1.75, 2.0]
C.void_number = 5
C.warm_up_epoch = 0

"""Eval Config"""
C.eval_epoch = 1
C.eval_stride_rate = 2 / 3
C.eval_scale_array = [1, ]  # 0.5, 0.75, 1, 1.5, 1.75
C.eval_flip = False
C.eval_base_size = 800
C.eval_crop_size = 800

"""Display Config"""
C.record_info_iter = 20
C.display_iter = 50

 Your project [work_space] name
C.proj_name = "OoD_Segmentation"

Your current experiment name
C.experiment_name = "pebal_baseline"

half pretrained_ckpts-loader upload images; loss upload every iteration
C.upload_image_step = [0, int((C.num_train_imgs / C.batch_size) / 2)]

False for debug; True for visualize
C.wandb_online = True

"""Save Config"""
C.saved_dir = os.path.join(C.root_dir, 'ckpts', C.experiment_name)

if not os.path.exists(C.saved_dir):
    os.mkdir(C.saved_dir)

Result:

**wandb:     Fishyscapes_ls_auprc 0.44404
wandb:     Fishyscapes_ls_auroc 0.98498
wandb:     Fishyscapes_ls_fpr95 0.06383
wandb: Fishyscapes_static_auprc 0.89534
wandb: Fishyscapes_static_auroc 0.9951
wandb: Fishyscapes_static_fpr95 0.0199
wandb:              energy_loss 0.03873
wandb:             gambler_loss 0.17864
wandb:              global_step 39**

**Comment by tianyu0207** [comment]
Hi 

> Hello , Thanks for your work
> 
> I followed the pebal config and trained using batch size 16 (as the paper) and 40 epochs However I cannot reproduce the paper result and check point result Is there any tips for reproducing the high AUPRC on the FS LS dataset?? Or is there any way to reproduce the result that has similar performance with checkpoint Could you help??
> 
> Thanks in advance Config : import os import numpy from easydict import EasyDict
> 
> C = EasyDict() config = C cfg = C
> 
> C.seed = 666
> 
> """Root Directory Config""" C.repo_name = 'pebal' C.root_dir = os.path.realpath(".")
> 
> """Data Dir and Weight Dir""" C.city_root_path = '/home/numb7315/PEBAL/code/dataset/city_scape' # path/to/your/city_scape C.coco_root_path = '/home/numb7315/PEBAL/code/dataset/coco' # path/to/your/coco C.fishy_root_path = '/home/numb7315/PEBAL/code/dataset/fishyscapes' # path/to/your/fishy
> 
> C.pebal_weight_path = os.path.join(C.root_dir, 'ckpts', 'pebal', 'best_ad_ckpt.pth') C.pretrained_weight_path = os.path.join(C.root_dir, 'ckpts', 'pretrained_ckpts', 'cityscapes_best.pth')
> 
> """Network Config""" C.fix_bias = True C.bn_eps = 1e-5 C.bn_momentum = 0.1
> 
> """Image Config""" C.num_classes = 19+1 # NOTE: 1 more channel for gambler loss C.image_mean = numpy.array([0.485, 0.456, 0.406]) # 0.485, 0.456, 0.406 C.image_std = numpy.array([0.229, 0.224, 0.225])
> 
> C.image_height = 900 C.image_width = 900
> 
> C.num_train_imgs = 2975 C.num_eval_imgs = 500
> 
> """Train Config""" C.lr = 1e-5 C.batch_size = 8 C.lr_power = 0.9 C.momentum = 0.9 C.weight_decay = 1e-4
> 
> C.nepochs = 40 C.niters_per_epoch = C.num_train_imgs // C.batch_size C.num_workers = 8 C.train_scale_array = [0.5, 0.75, 1, 1.5, 1.75, 2.0] C.void_number = 5 C.warm_up_epoch = 0
> 
> """Eval Config""" C.eval_epoch = 1 C.eval_stride_rate = 2 / 3 C.eval_scale_array = [1, ] # 0.5, 0.75, 1, 1.5, 1.75 C.eval_flip = False C.eval_base_size = 800 C.eval_crop_size = 800
> 
> """Display Config""" C.record_info_iter = 20 C.display_iter = 50
> 
> Your project [work_space] name C.proj_name = "OoD_Segmentation"
> 
> Your current experiment name C.experiment_name = "pebal_baseline"
> 
> half pretrained_ckpts-loader upload images; loss upload every iteration C.upload_image_step = [0, int((C.num_train_imgs / C.batch_size) / 2)]
> 
> False for debug; True for visualize C.wandb_online = True
> 
> """Save Config""" C.saved_dir = os.path.join(C.root_dir, 'ckpts', C.experiment_name)
> 
> if not os.path.exists(C.saved_dir): os.mkdir(C.saved_dir)
> 
> Result:
> 
> **wandb: Fishyscapes_ls_auprc 0.44404 wandb: Fishyscapes_ls_auroc 0.98498 wandb: Fishyscapes_ls_fpr95 0.06383 wandb: Fishyscapes_static_auprc 0.89534 wandb: Fishyscapes_static_auroc 0.9951 wandb: Fishyscapes_static_fpr95 0.0199 wandb: energy_loss 0.03873 wandb: gambler_loss 0.17864 wandb: global_step 39**

Hi, Thanks for your interest in our work. I also had a hard time selecting the best checkpoints across both FS LS and Static. It could potentially have different performances over different epochs in different machines. I would suggest tuning a bit of the hyper-parameters like learning rate and weights, or batch size. Or even the random state can have slightly different results as well. Sorry about the inconvenience.

**Comment by SnowdenLee** [comment]
Hi @hyunjunChhoi, 
I got similar results as yours. Did you manage to improve the performance further?  Thanks a lot!

@tianyu0207 May I ask if the current parameters in the config file are already the ones you used to get the best checkpoint? Thank you :)

**Comment by hyunjunChhoi** [comment]
> Hi @hyunjunChhoi, I got similar results as yours. Did you manage to improve the performance further? Thanks a lot!
> 
> @tianyu0207 May I ask if the current parameters in the config file are already the ones you used to get the best checkpoint? Thank you :)

Repeat the experiment multiple times. Then you can get a similar performance.
